enhancements/monitoring/alert-routing-for-user-workload-monitoring.md

```diff
@@ -0,0 +1,411 @@
+
+We already know from upstream users that Alertmanager can deal with many
+routes. We plan to simulate environments with thousands of `AlertmanagerConfig`
+resources and measure the impact on notification delivery.
+
+### Risks and Mitigations
+
+#### Disruption of the platform Alertmanager
+
+Even though the Prometheus operator prevents it as much as it can, it may be
+possible for users to create an `AlertmanagerConfig` resource that triggers the
+Prometheus operator to generate an invalid Alertmanager configuration, leading
+to a potential outage of the Platform Alertmanager cluster.
+
+Mitigations
+* The `AlertmanagerBadConfig` alert fires when Alertmanager can't reload its configuration.
+* Cluster admins can turn off the support for `AlertmanagerConfig` globally so that the Platform Alertmanager cluster can process platform alerts again and the cluster admins have time to identiy the "rogue" `AlertmanagerConfig` resource(s).
+* Cluster admins can exclude specific user namespaces (once the "rogue" `AlertmanagerConfig` resource(s) have been identified) to restore UWM functionality for good citizens.
+
+#### Misconfiguration of receivers
+
+Users may provide bad credentials for the receivers, the system receiving the
+notifications might be unreachable or the system might be unable to process the requests. These
+situations would trigger the `AlertmanagerFailedToSendAlerts` and/or
+`AlertmanagerClusterFailedToSendAlerts` alerts. The cluster admins have to act
+on upon the alerts and understand where the problem comes from.
+
+Mitigations
+* Detailed runbook for the `AlertmanagerFailedToSendAlerts` and `AlertmanagerClusterFailedToSendAlerts` alerts.
+* Ability to use a separate Alertmanager cluster to avoid messing up with the platform Alertmanager cluster.
+
+#### Non-optimal Alertmanager settings
+
+Users may use non-optimal settings for their alert notifications (such as
+reevaluation of alert groups at high frequency). This may impede the
+performances of Alertmanager globally because it would consume more CPU. It can
+also trigger notification failures if an exteral integration limits the number
+of requests a client IP address can do.
+
+Mitigation
+* Improve the `Alertmanager` CRD to expose new fields enforcing minimum interval values for all associated `AlertmanagerConfig` resources. This would be similar to what exists at the `Prometheus` CRD level for scrape targets with `enforcedSampleLimit` for instance.
+
+#### AlertmanagerConfig resources not being reconciled
+
+An `AlertmanagerConfig` resource might require credentials (such as API keys)
+which are referenced by secrets. If the Platform Prometheus operator doesn't
+have permissions to read the secret or if the reference is incorrect (wrong
+name or key), the operator doesn't reconcile the resource in the final
+Alertmanager configuration.
+
+Mitigation
+* The Prometheus operator should expose a validating admission webhook that should prevent invalid configurations.
```

Comment:
This would require that the webhook and the operator share the same RBAC to read cluster wide secrets, correct?

Reply:
the operator implements the webhook endpoint so it should have the proper RBAC but maybe it's better if the webhook doesn't try to resolve external references?
More a discussion for https://github.com/prometheus-operator/prometheus-operator/pull/4316 though ;)